src/main/java/fr/brouillard/oss/jgitver/impl/ScriptVersionStrategy.java

```diff
@@ -0,0 +1,309 @@
+import java.util.List;
+
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.lib.Ref;
+import org.eclipse.jgit.lib.Repository;
+import org.eclipse.jgit.revwalk.RevCommit;
+import org.eclipse.jgit.revwalk.RevWalk;
+import org.petitparser.context.Result;
+
+import org.apache.maven.shared.scriptinterpreter.BeanShellScriptInterpreter;
+import org.apache.maven.shared.scriptinterpreter.GroovyScriptInterpreter;
+import org.apache.maven.shared.scriptinterpreter.ScriptInterpreter;
+
+import fr.brouillard.oss.jgitver.Features;
+import fr.brouillard.oss.jgitver.ScriptType;
+import fr.brouillard.oss.jgitver.Version;
+
+import fr.brouillard.oss.jgitver.metadata.MetadataProvider;
+import fr.brouillard.oss.jgitver.metadata.MetadataRegistrar;
+import fr.brouillard.oss.jgitver.metadata.Metadatas;
+import fr.brouillard.oss.jgitver.metadata.TagType;
+
+/**
+ * Executes the given script (according its {@link ScriptType}, 
+ * default being {@link ScriptType#GROOVY}) 
+ * and use the script output as version.
+ *
+ * This output must be in CSV format using ';' separator, 
+ * with at least the 3 first columns, corresponding to integer values for
+ * major, minor and patch (e.g. <tt>1;2;3</tt>).
+ *
+ * The script has access to <a href="https://jgitver.github.io/#_meta_fields">metadata</a> (e.g. In BeanShell for the SHA1 <tt>metadata.GIT_SHA1_8</tt>).
+ *
+ * The script also has access to the environment variables 
+ * and system properties, respectively thanks to <tt>env</tt> 
+ * (e.g. <tt>env.HOME</tt>) and <tt>sys</tt> (e.g. <tt>sys.user.home</tt>).
+ *
+ * If no script is passed as parameter then the default one is used,
+ * with the same behavior as the <tt>PATTERN</tt> strategy.
+ */
+public class ScriptVersionStrategy extends VersionStrategy<ScriptVersionStrategy> {
+    private ScriptType scriptType = ScriptType.GROOVY;
+    private String script = "";
+
+    /**
+     * Default constructor.
+     *
+     * @param vnc        the configuration to use
+     * @param repository the git repository
+     * @param git        a git helper object built from the repository
+     * @param registrar  a storage for found/calculated metadata
+     */
+    public ScriptVersionStrategy(VersionNamingConfiguration vnc,
+                                 Repository repository,
+                                 Git git,
+                                 MetadataRegistrar registrar) {
+
+        super(vnc, repository, git, registrar);
+    }
+
+    @Override
+    public Version build(Commit head, List<Commit> parents) throws VersionCalculationException {
+        try {
+            final Commit base = findVersionCommit(head, parents);
+            final MetadataRegistrar registrar = getRegistrar();
+            final MetadataProvider metaProvider =
+                MetadataProvider.class.cast(registrar);
+
+            final HashMap<String, Object> metaProps =
+                new HashMap<String, Object>();
+
+            EnumSet.allOf(Metadatas.class).
```

Comment:
shouldn't this be done just before script execution so that you register only in the registrar and only once you copy the values from the registrar to the properties ?

Reply:
I don't think so, the value types are not the same in registrar and `metaProps` (e.g. `COMMIT_DISTANCE_TO_ROOT` exposed as integer in `metaProps`, not as string)